### ISSUE TITLE
bug in "bleachbit --sysinfo": unsupported os/platform functions in "/usr/share/bleachbit/bleachbit/Diagnostic.py"

### ISSUE THREAD

**Issue by verdy-p** [bleachbit/bleachbit#1355]
This version of Diagnostic.py updates the code to avoid this bug and correctly reporting system info with `os.uname()` and `platform`. Note that the code currently uses `platform.dist()` (which is not supported in any or most installation of Python, so it does not work in fact in Linux). It also attempts to use  `platform.linux_distribution()` by testing its existence, but it also does not exist in Python distributed by Debian, Ubuntu, Mint, etc. The code below is safer and adds support for documented Python API.

Note that this patches the current stable version 3.9.0-1 of bleachbit as it is distributed from Ubuntu "focal" repository.
(see "apt list bleachbit" to check that version, it should display "bleachbit/focal,focal,now 3.9.0-1 all  [installed]", and this version patches it to avoid the bug.

Note that more recent version 4.+ of Bleachbit are still NOT distributed by Ubuntu-based distribs, probably as it was considered unsafe and did not pass their security review, or there were too many unsolved reported bugs or even more compatiblity issues)

```
# vim: ts=4:sw=4:expandtab
# -*- coding: UTF-8 -*-

# BleachBit
# Copyright (C) 2008-2020 Andrew Ziem
# https://www.bleachbit.org
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


"""
Show diagnostic information
"""

import bleachbit

import locale
import os
import platform
import sys

if 'nt' == os.name:
    from win32com.shell import shell


def diagnostic_info():
    """Return diagnostic information as a string"""
    # this section is for application and library versions
    s = "BleachBit version %s" % bleachbit.APP_VERSION

    try:
        # Linux tarball will have a revision but not build_number
        from bleachbit.Revision import revision
        s += '\nGit revision %s' % revision
    except:
        pass
    try:
        from bleachbit.Revision import build_number
        s += '\nBuild number %s' % build_number
    except:
        pass
    try:
        import gi
        gi.require_version('Gtk', '3.0')
        from gi.repository import Gtk
        s += '\nGTK version {0}.{1}.{2}'.format(
            Gtk.get_major_version(), Gtk.get_minor_version(), Gtk.get_micro_version())
        s += '\nGTK theme = %s' % Gtk.Settings.get_default().get_property('gtk-theme-name')
        s += '\nGTK icon theme = %s' % Gtk.Settings.get_default().get_property('gtk-icon-theme-name')
        s += '\nGTK prefer dark theme = %s' % Gtk.Settings.get_default().get_property('gtk-application-prefer-dark-theme')
    except:
        pass
    import sqlite3
    s += "\nSQLite version %s" % sqlite3.sqlite_version

    # this section is for variables defined in __init__.py
    s += "\nlocal_cleaners_dir = %s" % bleachbit.local_cleaners_dir
    s += "\nlocale_dir = %s" % bleachbit.locale_dir
    s += "\noptions_dir = %s" % bleachbit.options_dir
    s += "\npersonal_cleaners_dir = %s" % bleachbit.personal_cleaners_dir
    s += "\nsystem_cleaners_dir = %s" % bleachbit.system_cleaners_dir

    # this section is for information about the system environment
    s += "\nlocale.getdefaultlocale = %s" % str(locale.getdefaultlocale())
    if 'posix' == os.name:
        envs = ('DESKTOP_SESSION', 'LOGNAME', 'USER', 'SUDO_UID')
    if 'nt' == os.name:
        envs = ('APPDATA', 'cd', 'LocalAppData', 'LocalAppDataLow', 'Music',
                'USERPROFILE', 'ProgramFiles', 'ProgramW6432', 'TMP')
    for env in envs:
        s += "\nos.getenv('%s') = %s" % (env, os.getenv(env))
    s += "\nos.path.expanduser('~') = %s" % os.path.expanduser('~')

    if hasattr(os, 'uname'):
        if hasattr(os.uname(), 'sysname'):
            s += "\nos.uname().sysname = %s" % os.uname().sysname
        if hasattr(os.uname(), 'nodename'):
            s += "\nos.uname().nodename = %s" % os.uname().nodename
        if hasattr(os.uname(), 'release'):
            s += "\nos.uname().release = %s" % os.uname().release
        if hasattr(os.uname(), 'version'):
            s += "\nos.uname().version = %s" % os.uname().version
        if hasattr(os.uname(), 'machine'):
            s += "\nos.uname().machine = %s" % os.uname().machine
    if hasattr(platform, 'platform'):
        s += "\nplatform.platform = %s" % platform.platform()
    if hasattr(platform, 'release'):
        s += "\nplatform.release = %s" % str(platform.release())
    if hasattr(platform, 'version'):
        s += "\nplatform.version = %s" % platform.version()
    if hasattr(platform, 'uname'):
        s += "\nplatform.uname = %s" % str(platform.uname())

    if sys.platform.startswith('linux'):
        if hasattr(platform, 'linux_distribution'):
            s += "\nplatform.linux_distribution() = %s" % str(
                platform.linux_distribution())
        elif hasattr(platform, 'dist'):
            s += "\nplatform.dist() = %s" % str(platform.dist())

    # Mac Version Name - Dictionary
    macosx_dict = {'5': 'Leopard', '6': 'Snow Leopard', '7': 'Lion', '8': 'Mountain Lion',
                   '9': 'Mavericks', '10': 'Yosemite', '11': 'El Capitan', '12': 'Sierra'}
    if sys.platform.startswith('darwin'):
        if hasattr(platform, 'mac_ver'):
            for key in macosx_dict:
                if (platform.mac_ver()[0].split('.')[1] == key):
                    s += "\nplatform.mac_ver() = %s" % str(
                        platform.mac_ver()[0] + " (" + macosx_dict[key] + ")")
        else:
            s += "\nplatform.dist() = %s" % str(platform.dist())

    if 'nt' == os.name:
        s += "\nplatform.win32_ver[1]() = %s" % platform.win32_ver()[1]

    s += "\nsys.argv = %s" % sys.argv
    s += "\nsys.executable = %s" % sys.executable
    s += "\nsys.version = %s" % sys.version
    if 'nt' == os.name:
        s += "\nwin32com.shell.shell.IsUserAnAdmin() = %s" % shell.IsUserAnAdmin(
        )
    s += "\n__file__ = %s" % __file__

    return s

```

**Comment by az0** [comment]
1. I think this issue was already fixed here in BleachBit 4.0.0 https://github.com/bleachbit/bleachbit/commit/282de2198914a3d24b43a37e31160f6f132d7865

2. New versions of Ubuntu have new versions of BleachBit. For example, Jammy (22.04) has BleachBit 4.4.2 https://launchpad.net/ubuntu/jammy/+source/bleachbit . It's always been this way. For a given version of Ubuntu, there were never any BleachBit updates because of Ubuntu's no rolling release policy, so I encourage Ubuntu users to download BleachBit from https://www.bleachbit.org .

3. Please use pull requests to submit code changes

**Comment by abitrolly** [comment]
@verdy-p you should report it to Ubuntu as https://wiki.ubuntu.com/StableReleaseUpdates#High-impact_bugs so they can update to newer version.

**Comment by az0** [comment]
BleachBit version 3.9 is not supported, and users should please update to the latest version. The decision for which version of BleachBit (and which patches are applied) is up to the Ubuntu maintainers. If I hypothetically were to accept a patch to fix a bug for 3.9, it wouldn't affect the repositories of old Ubuntu repositories for the same reason: the Ubuntu policy. Even if Ubuntu maintainers are not going to change it, Ubuntu users can get the latest version as a .deb installer from https://www.bleachbit.org/download .
